designs/move-cluster-specific-code-out-of-manager.md

```diff
@@ -0,0 +1,209 @@
+	GetFieldIndexer() client.FieldIndexer
+
+	// GetCache returns a cache.Cache
+	GetCache() cache.Cache
+
+	// GetEventRecorderFor returns a new EventRecorder for the provided name
+	GetEventRecorderFor(name string) record.EventRecorder
+
+	// GetRESTMapper returns a RESTMapper
+	GetRESTMapper() meta.RESTMapper
+
+	// GetAPIReader returns a reader that will be configured to use the API server.
+	// This should be used sparingly and only when the client does not fit your
+	// use case.
+	GetAPIReader() client.Reader
+
+	// GetScheme returns an initialized Scheme
+	GetScheme() *runtime.Scheme
+
+	// Start starts the ClusterConnector
+	Start(<-chan struct{}) error
+}
+```
+
+And the current `Manager` interface will change to look like this:
+
+```go
+type Manager interface {
+	// ClusterConnector holds objects to connect to a cluster
+	cluserconnector.ClusterConnector
+
+	// Add will set requested dependencies on the component, and cause the component to be
+	// started when Start is called.  Add will inject any dependencies for which the argument
+	// implements the inject interface - e.g. inject.Client.
+	// Depending on if a Runnable implements LeaderElectionRunnable interface, a Runnable can be run in either
+	// non-leaderelection mode (always running) or leader election mode (managed by leader election if enabled).
+	Add(Runnable) error
+
+	// Elected is closed when this manager is elected leader of a group of
+	// managers, either because it won a leader election or because no leader
+	// election was configured.
+	Elected() <-chan struct{}
+
+	// SetFields will set any dependencies on an object for which the object has implemented the inject
+	// interface - e.g. inject.Client.
+	SetFields(interface{}) error
+
+	// AddMetricsExtraHandler adds an extra handler served on path to the http server that serves metrics.
+	// Might be useful to register some diagnostic endpoints e.g. pprof. Note that these endpoints meant to be
+	// sensitive and shouldn't be exposed publicly.
+	// If the simple path -> handler mapping offered here is not enough, a new http server/listener should be added as
+	// Runnable to the manager via Add method.
+	AddMetricsExtraHandler(path string, handler http.Handler) error
+
+	// AddHealthzCheck allows you to add Healthz checker
+	AddHealthzCheck(name string, check healthz.Checker) error
+
+	// AddReadyzCheck allows you to add Readyz checker
+	AddReadyzCheck(name string, check healthz.Checker) error
+
+	// Start starts all registered Controllers and blocks until the Stop channel is closed.
+	// Returns an error if there is an error starting any controller.
+	// If LeaderElection is used, the binary must be exited immediately after this returns,
+	// otherwise components that need leader election might continue to run after the leader
+	// lock was lost.
+	Start(<-chan struct{}) error
+
+	// GetWebhookServer returns a webhook.Server
+	GetWebhookServer() *webhook.Server
+}
+```
+
```

Comment:
Would it be possible to separate this into 3 interfaces?

```go
type cluserconnector.ClusterConnector{
    // same as above
...
}}
type MultiClusterManager interface {
    // Create a new connection to a cluster. Creates a new instance of ClusterConnector and adds it to this manager
    Connect(config *rest.Config, name string, opts ...Option) cluserconnector.ClusterConnector
    // same as Manager from proposal except cluserconnector.ClusterConnector
    Add(Runnable) error
    Elected() <-chan struct{}
    SetFields(interface{}) error
    AddMetricsExtraHandler(path string, handler http.Handler) error
    AddHealthzCheck(name string, check healthz.Checker) error
    AddReadyzCheck(name string, check healthz.Checker) error
    Start(<-chan struct{}) error
    GetWebhookServer() *webhook.Server
}

type Manager interface {
     cluserconnector.ClusterConnector
     MultiClusterManager
}
```

Introducing a factory method `Connect` inside `MultiClusterManager` would also solve the problems a having to use type assertions.

Reply:
@kramerul Generally I like the idea (although I would probably embedd the `Manager` ito the `MultiClusterManager` and not the other way round) but it opens up an interesting question: What config will we use for leader election?

Reply:
@alvaroaleman, I haven't thought about this aspect. But it's absolutely correct, that you need to have a primary connection for leader election. In this case I would agree to have only one `Manager` interface.

The `Connect` method could ease the usage of the API.

Reply:
I like the idea of having an explicit interface for MultiCluster manager.

> I would probably embedd the Manager ito the MultiClusterManager and not the other way round)

+1

A possible alternative is:

```
type ClusterConnector{
    // same as above
}

// For single cluster use-case only
type Manager interface {
     // same as above
}

type MultiClusterManager interface {
    Manager

    // Create a new connection to a secondary cluster. Creates a new instance of ClusterConnector and adds it to this manager. Each ClusterConnector should have its unique name.
    ConnectSecondaryCluster(config *rest.Config, name string, opts ...Option) (cluserconnector.ClusterConnector, error)
    // Look up the ClusterConnector by its unique name.
    GetClusterConnectorFor(name string) (cluserconnector.ClusterConnector, error)
}
```

Reply:
I've added it like this:
```
type MultiClusterManager interface {
       Manager
       // Add another named cluster to the MultiClusterManager. The name
       // must be unique. The MultiClusterManager will wait for all clusters
       // caches to be started before starting anytything else.
       AddCluster(config *rest.Config, name string) error
       // GetClusters returns all Clusters this MultiClusterManager knows about.
       // The cluster used to construct the MultiClusterManager is named `primary`
       // and will be used for LeaderElection, if enabled.
       GetClusters() map[string]clusterconnector.ClusterConnector
}
```

In all the multi-cluster controllers I've built so far, the actual controller just wants to watch in all clusters but doesn't necessarily know the number or names of them ahead of time which is why a method that returns a map of all clusters is IMHO more useful. If there is one that gets a special treatment, this is still possible.

Reply:
A number of the patterns I've seen for multicluster end up having clusters come & go over time.  Even if we're not going to tackle that now, we should keep that in mind while designing.

I'm also not certain about the whole "primary" / "secondary" cluster thing.  Perhaps modeling this as `there's a "leader election" cluster and "functional" clusters (one of which may be the leader election cluster)`?

Reply:
> A number of the patterns I've seen for multicluster end up having clusters come & go over time. Even if we're not going to tackle that now, we should keep that in mind while designing.

IMHO this proposal should do nothing to prevent that but it does not need to do anything yet to simplify that. Once we want such a functionality, we would probably add a `RemoveCluster` to the interface.

> I'm also not certain about the whole "primary" / "secondary" cluster thing. Perhaps modeling this as there's a "leader election" cluster and "functional" clusters (one of which may be the leader election cluster)?

I guess we could probably change `AddCluster` to have a `UseForLeaderElection` bool opt if leader election is enabled, use that cluster (and error out when starting and note exactly one cluster is configured to be used for leader election). WDYT?

Reply:
On second thought, after trying to update the sample, I would like to not make a `MultiClusterManager` part of this proposal, because it opens up a set of questions I would like to not answer just yet:

* The `Builder` needs to be extended to have something like `WatchesFromCluster(source, "cluster-name"`, handler)`, otherwise we need to blindly dereference elements from the `map[clusterName]cluster.Cluster`, potentially getting NPDs
* The `Builder` needs to be extended to have something like `WatchesFromAllClustersExcept(source, "exepcted-cluster-name", handler) 
* Probably more use cases for watches I didn't think about
* There should be a nice way to pass on all clients for all clusters to a controller that doesn't involve writing a loop (probably the multi cluster client abstraction that Solly suggested?)
* When adding official support for building a controller that watches an arbitraty number of clusters, we also need an official way of encoding the cluster name into the `reconcile.Request`

And probably more.

IMHO, extracting the clusterspecifics out of the manager won't block any of the work mentioned above but allows us to start working on the topic without requiring us to already anticipate all use cases.

Reply:
ok, fine with doing that in a follow-up